sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -537,6 +537,7 @@ public ProcessContinuation processElement(
       while (tracker.tryClaim(out) && out[0] != null) {
         receiver.outputWithTimestamp(
             new ValueWithRecordId<>(out[0].getValue(), out[0].getId()), out[0].getTimestamp());
+        watermarkEstimator.setWatermark(tracker.currentRestriction().getWatermark());
```

Comment:
This feels a little expensive. Underneath it calls UnboundedReader#getCheckpointMark, which is supposed to be called (mostly) before checkpoint, not for each element.

Reply:
It seems to be possible to do this by passing the watermark in `UnboundedSourceValue`, though.

Reply:
Hm, actually ... this seems to be a no-op, because the watermark estimator used is WatermarkEstimators.Manual, and that simply overwrites the set watermark in `lastReportedWatermark`. The watermark is then read only when the `processElement` terminates. Therefore, I think we should leave the call to `setWatemark` only outside of this while loop.

Reply:
>  I think we should leave the call to setWatemark only outside of this while loop.

The tricky part is that when we exit `tryClaim()` loop, it's either trySplit() has been called or when source reader return false from advance(). When trySplit() is called, the currentRestriction will be changed to `EmptyUnboundedSource `. So output of the `tryClaim()` loo, the watermark we get from `currentRestriction` is actually from `EmptyUnboundedSource`, which is MAX_TIMESTAMP.

Besides, the SDK gets watermark for `DelayedBundleApplication` from `WatermarkEstimator` before we call trySplit[`]. So we actually need to update the WatermarkEstimator with correct watermark every time we do tryClaim.
[1] https://github.com/apache/beam/blob/3bb232fb098700de408f574585dfe74bbaff7230/runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java#L330-L331